cloudinit/sources/DataSourceAzure.py

```diff
@@ -719,30 +727,48 @@ def get_public_ssh_keys(self):
         advantage, so this is a strong preference. But we must keep
         OVF as a second option for environments that don't have IMDS.
         """
+
         LOG.debug('Retrieving public SSH keys')
         ssh_keys = []
+        keys_from_imds = True
         try:
-            raise KeyError(
-                "Not using public SSH keys from IMDS"
-            )
-            # pylint:disable=unreachable
             ssh_keys = [
                 public_key['keyData']
                 for public_key
                 in self.metadata['imds']['compute']['publicKeys']
             ]
-            LOG.debug('Retrieved SSH keys from IMDS')
+            for key in ssh_keys:
+                if not _key_is_openssh_formatted(key=key):
+                    keys_from_imds = False
+                    break
+
+            if not keys_from_imds:
+                log_msg = 'Keys not in OpenSSH format, using OVF'
+            else:
+                log_msg = 'Retrieved {} keys from IMDS'.format(
+                    len(ssh_keys)
+                    if ssh_keys is not None
+                    else 0
+                )
         except KeyError:
             log_msg = 'Unable to get keys from IMDS, falling back to OVF'
+            keys_from_imds = False
+        finally:
             report_diagnostic_event(log_msg, logger_func=LOG.debug)
+
+        if not keys_from_imds:
             try:
                 ssh_keys = self.metadata['public-keys']
-                LOG.debug('Retrieved keys from OVF')
+                log_msg = 'Retrieved {} keys from OVF'.format(len(ssh_keys))
             except KeyError:
                 log_msg = 'No keys available from OVF'
+            finally:
                 report_diagnostic_event(log_msg, logger_func=LOG.debug)
 
-        return ssh_keys
+        return dict(
```

Comment:
A nit, but using a dict here feels a little off: this is really an immutable 2-tuple. I appreciate that using a dict allows us to use names (instead of indexing) for these: a more Pythonic way of achieving the same would be to use a `namedtuple`:

```py
In [8]: keys_from_imds = False

In [9]: ssh_keys = ["ssh-rsa definitelyavalidkey", "ssh-dsa thisonesbadthough"]

In [10]: return_value = namedtuple("AzureSSHKeys", ("keys_from_imds", "ssh_keys"))(keys_from_imds, ssh_keys)

In [11]: return_value.keys_from_imds
Out[11]: False

In [12]: return_value.ssh_keys
Out[12]: ['ssh-rsa definitelyavalidkey', 'ssh-dsa thisonesbadthough']
```

The namedtuple definition should probably go at module level, though.

(You could also consider returning a `(2, tuple)` and consistently deconstructing it with `keys_from_imds, ssh_keys = self._get_public_ssh_keys_and_source()`.)

Reply:
Good idea! Change made (and tested).